src/middleware/mod.rs

```diff
@@ -40,3 +49,99 @@ pub trait Middleware<Data>: Send + Sync {
     ) -> FutureObj<'static, ()>;
     */
 }
+
+/// Middleware that wraps around remaining middleware chain.
+pub trait Middleware<Data>: Send + Sync {
+    /// Asynchronously handle the request, and return a response.
+    fn handle<'a>(&'a self, ctx: RequestContext<'a, Data>) -> FutureObj<'a, ResponseContext<Data>>;
+}
+
+impl<Data, F> Middleware<Data> for F
+where
+    F: Send + Sync + Fn(RequestContext<Data>) -> FutureObj<ResponseContext<Data>>,
+{
+    fn handle<'a>(&'a self, ctx: RequestContext<'a, Data>) -> FutureObj<'a, ResponseContext<Data>> {
+        (self)(ctx)
+    }
+}
+
+pub struct ReqResAdapter<T>(T);
+
+impl<T, Data> Middleware<Data> for ReqResAdapter<T>
+where
+    T: ReqResMiddleware<Data>,
+    Data: Clone + Send,
+{
+    fn handle<'a>(
+        &'a self,
+        mut ctx: RequestContext<'a, Data>,
+    ) -> FutureObj<'a, ResponseContext<Data>> {
+        let m = &self.0;
+        FutureObj::new(Box::new(
+            async move {
+                if let Err(res) = await!(m.request(&mut ctx.app_data, &mut ctx.req, &ctx.params)) {
+                    return ctx.respond(res);
+                }
+                let mut ctx = await!(ctx.next());
+                await!(m.response(&mut ctx.app_data, &ctx.head, &mut ctx.res));
+                ctx
+            },
+        ))
+    }
+}
+
+pub struct RequestContext<'a, Data> {
+    pub app_data: Data,
+    pub req: Request,
+    pub params: RouteMatch<'a>,
+    pub(crate) endpoint: &'a BoxedEndpoint<Data>,
+    pub(crate) next_middleware: &'a [Arc<dyn Middleware<Data> + Send + Sync>],
+}
+
+impl<'a, Data> RequestContext<'a, Data> {
+    /// Consume this context, and respond with given `Response`.
+    pub fn respond(self, res: Response) -> ResponseContext<Data> {
+        ResponseContext::from_request_response(self, res)
+    }
+}
+
+impl<'a, Data: Clone + Send> RequestContext<'a, Data> {
+    /// Consume this context, and run remaining middleware chain to completion.
+    pub fn next(mut self) -> FutureObj<'a, ResponseContext<Data>> {
+        FutureObj::new(Box::new(
+            async move {
+                if let Some((current, next)) = self.next_middleware.split_first() {
+                    self.next_middleware = next;
+                    await!(current.handle(self))
+                } else {
+                    let (head, res) =
+                        await!(self
+                            .endpoint
+                            .call(self.app_data.clone(), self.req, self.params));
+                    ResponseContext {
+                        app_data: self.app_data,
+                        res,
+                        head,
+                    }
+                }
+            },
+        ))
+    }
+}
+
+pub struct ResponseContext<Data> {
```

Comment:
So, here's an interesting thing: the main reason to have the `Head` type at all (and to return it from the endpoint) was to support "after"-style middleware getting access to most of the request. But with around middleware, we can maybe simplify.

Because of compiler limitations, we aren't currently able to have endpoints get `&Request` but have to pass full ownership. Eventually we'll be able to pass in a borrow.

But with around middleware, if you want to do something with request data after the endpoint runs, you can just compute what you need *before* calling `next`, and then refer to it later.

----

I'm thinking we just drop `head`, `Head`, and `ResponseContext` altogether, and just return a `Response`. This would also let us resolve https://github.com/rust-net-web/tide/issues/41#issuecomment-438496200

What do you think?

Reply:
I've totally missed that point! We can definitely simplify this.